### ISSUE TITLE
Set per-feed curl options

### ISSUE THREAD

**Issue by sanel** [skeeto/elfeed#508]
Is there a way to have per-feed a specific curl args set? I've noticed that some sites, with `--compressed` argument and http2, will return `curl: (23) Failed writing received data to disk/application`, probably due to [non-conforming web server](https://github.com/curl/curl/issues/4092). For example:

```sh
# works
$ curl --compressed https://news.ycombinator.com/rss >/dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 11404    0 11404    0     0  14087      0 --:--:-- --:--:-- --:--:-- 14079

# fails, but looks like all content is downloaded
$ curl --compressed https://www.klix.ba/rss >/dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  7193    0  7193    0     0  42774      0 --:--:-- --:--:-- --:--:-- 42815
curl: (23) Failed writing received data to disk/application

# but this works
$ curl --compressed --http1.1 https://www.klix.ba/rss >/dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  7193    0  7193    0     0  44618      0 --:--:-- --:--:-- --:--:-- 44401
```

curl version where this happens:

```
curl 8.7.1 (x86_64-pc-linux-gnu) libcurl/8.7.1 OpenSSL/1.1.1w zlib/1.2.13 brotli/1.0.9 zstd/1.5.5 c-ares/1.19.1 libidn2/2.3.2 libpsl/0.21.1 libssh2/1.10.0 nghttp2/1.57.0 OpenLDAP/2.4.59
Release-Date: 2024-03-27
Protocols: dict file ftp ftps gopher gophers http https imap imaps ipfs ipns ldap ldaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM PSL SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd
~/programs/projects/stylitics/ansible-playbooks $ 
```

but it works with older curl version, like:

```
curl 7.81.0 (x86_64-pc-linux-gnu) libcurl/7.81.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.16
Release-Date: 2022-01-05
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets zstd
```

Funny thing is that Emacs 29.3 url package works without problems, with `(setq elfeed-use-curl nil)` set.

**Comment by minad** [comment]
I cannot reproduce this problem anymore with curl 8.14.1 which is available on Debian stable. My opinion is that we should not add per-feed curl options if the goal is only to work around bugs in curl or some web servers. My suggestion would be to advise `elfeed-curl--args` to change or remove problematic settings, which affect your usage. If we identify certain settings as generally problematic, e.g., `--compressed` we could also make it globally customizable.

**Comment by sanel** [comment]
You suggest to advise `(elfeed-curl--args)` per feed url? How? I think having per-feed curl settings is beneficial, because e.g. you'll be able to emit site specific stuff/headers like http auth, custom certificate or older TLS version, which can be common in government sites, that are not always up-to-date.

**Comment by minad** [comment]
As I wrote, I could not reproduce the problem with the URLs you gave above. But if some command line option turns out to be highly problematic, you could remove it for all curl invocations by advising `elfeed-curl--args`. If it becomes clear that more users are affected by a problematic command line option, we could either remove the option completely, or introduce a customization option e.g., `elfeed-curl-default-arguments`.

I don't think we should offer per-feed curl options - this feels too low level. For sites which need authentication you could try to use a netrc file and configure it via `elfeed-curl-extra-arguments`. But if there are other good reasons for per-feed curl options, then we can reconsider. However it does not look as if many users have requested such a feature before.

**Comment by yantar92** [comment]
Daniel Mendler ***@***.***> writes:

> ... My opinion is that we should not add per-feed curl options if the
> goal is only to work around bugs in curl or some web servers

+1

-- 
Ihor Radchenko // yantar92,
Org mode maintainer,
Learn more about Org mode at <https://orgmode.org/>.
Support Org development at <https://liberapay.com/org-mode>,
or support my work at <https://liberapay.com/yantar92>

**Comment by minad** [comment]
@yantar92 As I've understood @sanel now the issue seems to be broader, which wasn't clear from the initial report (only about curl options due to curl bugs/incompatibilities). There might be scenarios where we'd still want such configurability, but I wouldn't add it lightheartedly. Since this hasn't been requested often, I think we can wait and see.

**Comment by sanel** [comment]
@minad Correct! I’ve encountered several cases where per-feed curl options would have been highly beneficial:

1. Some endpoints only worked with http 1.1, while others used http 2.0 (curl's default). One example was reported here, though it was eventually fixed by the site owners.
2. One site was restricted to specific regions, requiring a socks proxy. Because I couldn't set this per-feed, I had to route all my feeds through proxy, which slowed everything down.
3. Dfferent sites sometimes require different user-agents to load properly.

There are a few other use cases as well, but I can't recall them at the moment. I'll investigate if I can address this by using some `advice-add` workaround in my config. Thanks for the feedback!
